File: C++/tower.cpp
// Stuart Barclay 07/01/2019

#include "tower.h"

tower::tower(int poleCount, int nodeCount) : poleCount(poleCount)
{
    this->moveCount = 0;
    // ASCII for 'A'
    char name = 65;
    bool empty = false;
    this->poles = new towerPole *[poleCount];
    this->poles[0] = new towerPole(nodeCount, empty, name++);
    empty = true;
    for (int i = 1; i < poleCount; i++)
        this->poles[i] = new towerPole(nodeCount, empty, name++);
}

// Creates a 2D array to build the tower and then deletes the array when complete
void tower::printTower()
{
    cout << "________________________________________________________________________________\n\n";
    int **tower = new int *[this->poleCount];
    int nodes;
    poleNode *currentNode;
    for (int i = 0; i < poleCount; i++)
    {
        nodes = this->poles[i]->getNodeCount();
        tower[i] = new int[nodes];
        currentNode = this->poles[i]->getPole();
        for (int j = 0; j < nodes; j++)
        {
            tower[i][j] = currentNode->getValue();
            currentNode = currentNode->nextNode;
        }
    }

    for (int i = 0; i < nodes; i++)
    {
        cout << "\t";
        for (int j = 0; j < poleCount; j++)
        {

            if (tower[j][i] == -1)
                cout << "|";
            else
                cout << tower[j][i];
            cout << "\t\t";
        }
        cout << endl;
    }

    cout << "\t";
    for (int j = 0; j < poleCount; j++)
        cout << "_\t\t";
    cout << endl
         << "\t";
    for (int j = 0; j < poleCount; j++)
        cout << this->poles[j]->getPoleName() << "\t\t";
    cout << endl;

    for (int i = 0; i < this->poleCount; ++i)
    {
        delete[] tower[i];
    }
    delete[] tower;
}

tower::~tower()
{
    for (int i = 0; i < poleCount; i++)
    {
        if (this->poles[i] != nullptr)
        {
            this->poles[i]->~towerPole();
            delete poles[i];
        }
    }
    delete[] poles;
    poles = nullptr;
}

void tower::playGame()
{
    this->moveCount = 0;
    while (!gameFinished())
    {
        char poleMoveFrom, poleMoveTo;
        cout << "Which pole would you like to move the top disk from (A - " << this->poles[this->poleCount - 1]->getPoleName() << ")? ";
        cin >> poleMoveFrom;
        cout << "Which pole would you like to move the top disk to (A - " << this->poles[this->poleCount - 1]->getPoleName() << ")? ";
        cin >> poleMoveTo;
        if (!this->poles[toupper(poleMoveFrom) - 65]->reallocatePole(this->poles[(toupper(poleMoveTo) - 65)]))
        {
            cout << "Invalid move!\n";
            this->printTower();
            continue;
        }
        ++this->moveCount;
        cout << "\n";
        this->printTower();
    }
    cout << "________________________________________________________________________________\n\t\tCongratulations! Game completed in " << this->moveCount << " moves.\n________________________________________________________________________________";
}

bool tower::gameFinished()
{
    poleNode *lastPole = this->poles[this->poleCount - 1]->getPole();
    while (lastPole)
    {
        if (lastPole->getValue() == -1)
            return false;
        lastPole = lastPole->nextNode;
    }
    return true;
}

File: Java/src/tower.java
import java.util.LinkedList;
import java.util.Scanner;

public class tower {
    private int poleCount;
    private LinkedList<towerPole> poles;
    private int moveCount;

    public tower(int poleCount, int nodeCount) {
        this.poles = new LinkedList<towerPole>();
        this.setPoleCount(poleCount);
        this.setMoveCount(0);
        char name = 65;
        this.addPoles(new towerPole(nodeCount, false, name++));
        for (int i = 1; i < poleCount; ++i)
            this.addPoles(new towerPole(nodeCount, true, name++));
    }

    public void printTower() {
        print("________________________________________________________________________________\n");
        int[][] tower = new int[this.getPoleCount()][this.getPoles().peek().getNodeCount()];
        int poleCounter = 0, nodeCounter = 0;
        for (towerPole Pole : this.getPoles()) {
            poleNode pNode = Pole.getPoleHead();
            tower[poleCounter] = new int[Pole.getNodeCount()];
            while (pNode != null) {
                tower[poleCounter][nodeCounter++] = pNode.getValue();
                pNode = pNode.getNextNode();
            }
            ++poleCounter;
            nodeCounter = 0;
        }
        poleCounter = 0;
        nodeCounter = 0;
        for (int i = 0; i < this.getPoles().peek().getNodeCount(); ++i) {
            print("\t");
            for (int j = 0; j < this.getPoleCount(); ++j) {
                if (tower[j][i] == -1)
                    print("|");
                else
                    print(Integer.toString(tower[j][i]));
                print("\t\t");
            }
            println("");
        }
        print("\t");

        for (int i = 0; i < this.getPoleCount(); ++i)
            print("_\t\t");

        println("");
        print("\t");
        for (towerPole p : this.getPoles())
            print(String.valueOf(p.getPoleName()) + "\t\t");
        println("");
    }

    public void playGame() {
        Scanner scanner = new Scanner(System.in);
        while(!this.gameFinished()){
            char poleMoveFrom, poleMoveTo;
            println("Which pole would you like to move the top disk from (A - " + this.getPoles().getLast().getPoleName() + ") ?");
            poleMoveFrom = scanner.nextLine().toUpperCase().charAt(0);
            println("Which pole would you like to move the top disk to (A - " + this.getPoles().getLast().getPoleName() + ") ?");
            poleMoveTo = scanner.nextLine().toUpperCase().charAt(0);
            if(!this.getPoles().get(poleMoveFrom - 65).reallocatePole(this.getPoles().get(poleMoveTo - 65))){
                println("Invald move!");
                this.printTower();
                continue;
            }
            this.setMoveCount(this.getMoveCount()+ 1);
            println("");
            this.printTower();
        }
        println("________________________________________________________________________________\n\t\tCongratulations! Game completed in " + Integer.toString(this.getMoveCount()) + " moves.\n________________________________________________________________________________");
    }

    public boolean gameFinished() {
        poleNode lastPole = this.getPoles().getLast().getPoleHead();
        while (lastPole != null) {
            if (lastPole.getValue() == -1)
                return false;
            lastPole = lastPole.getNextNode();
        }
        return true;
    }

    public void addPoles(towerPole pole) {
        this.poles.add(pole);
    }

    public void print(String printString) {
        System.out.print(printString);
    }

    public void println(String printString) {
        System.out.println(printString);
    }

    public int getPoleCount() {
        return poleCount;
    }

    public void setPoleCount(int poleCount) {
        this.poleCount = poleCount;
    }

    public LinkedList<towerPole> getPoles() {
        return poles;
    }

    public void setPoles(LinkedList<towerPole> poles) {
        this.poles = poles;
    }

    public int getMoveCount() {
        return moveCount;
    }

    public void setMoveCount(int moveCount) {
        this.moveCount = moveCount;
    }
}

